jetty-websocket/javax-websocket-server-impl/src/main/java/org/eclipse/jetty/websocket/jsr356/server/deploy/WebSocketServerContainerInitializer.java

```diff
@@ -129,61 +131,145 @@ public static boolean isEnabledViaContext(ServletContext context, String keyName
-                if (LOG.isDebugEnabled())
-                    LOG.debug("Dynamic filter add to support JSR356/javax.websocket.server: {}", WebSocketUpgradeFilter.class.getName());
-                WebSocketUpgradeFilter wsuf = WebSocketUpgradeFilter.configureContext(context);
-                context.setAttribute(instanceKey, wsuf);
+                String instanceKey = WebSocketUpgradeFilter.class.getName() + ".SCI";
+                if(context.getAttribute(instanceKey) == null)
+                {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("Dynamic filter add to support JSR356/javax.websocket.server: {}", WebSocketUpgradeFilter.class.getName());
+                    WebSocketUpgradeFilter wsuf = WebSocketUpgradeFilter.configureContext(servletContextHandler);
+                    context.setAttribute(instanceKey, wsuf);
+                }
             }
         }
-    
-        return jettyContainer;
     }
-    
+
     /**
-     * @deprecated use {@link #configureContext(ServletContextHandler)} instead
-     * @param context not used
-     * @param jettyContext the {@link ServletContextHandler} to use
-     * @return a configured {@link ServerContainer} instance
-     * @throws ServletException if the {@link WebSocketUpgradeFilter} cannot be configured
+     * Configure the {@link ServletContextHandler} to call {@link WebSocketServerContainerInitializer#onStartup(Set, ServletContext)}
+     * during the {@link ServletContext} initialization phase.
+     *
+     * @param context the context to add listener to
      */
-    @Deprecated
-    public static ServerContainer configureContext(ServletContext context, ServletContextHandler jettyContext) throws ServletException
+    public static void configure(ServletContextHandler context)
     {
-        return configureContext(jettyContext);
+        context.addEventListener(ContainerInitializer.asContextListener(new WebSocketServerContainerInitializer()));
     }
 
+    /**
+     * Configure the {@link ServletContextHandler} to call {@link WebSocketServerContainerInitializer#onStartup(Set, ServletContext)}
+     * during the {@link ServletContext} initialization phase.
+     *
+     * @param context the context to add listener to
+     * @param configurator the lambda that is called to allow the {@link ServerContainer} to
+     * be configured during the {@link ServletContext} initialization phase
+     */
+    public static void configure(ServletContextHandler context, Configurator configurator)
+    {
+        // In this embedded-jetty usage, allow ServletContext.addListener() to
+        // add other ServletContextListeners (such as the ContextDestroyListener) after
+        // the initialization phase is over. (important for this SCI to function)
+        context.getServletContext().setExtendedListenerTypes(true);
+
+        context.addEventListener(
+            ContainerInitializer.asContextListener(new WebSocketServerContainerInitializer())
```

Comment:
why do we need both `Configurator` and a `initConsumer`?   Why can't the `ContainerInitizer` just know about the `Configurator` and call it directly?  Feels like two levels of abstraction when one could do.

Reply:
thinking about this ...

Reply:
`initConsumer` is generic for the `ContainerInitializer`.

We have 2 `Configurators`, which are specific to the SCI implementations. (the `BiConsumer` usage made this difference obvious)

* NativeWebSocketServletContainerInitializer.Configurator has `void accept(ServletContext servletContext, NativeWebSocketConfiguration nativeWebSocketConfiguration);`
* WebSocketServerContainerInitializer.Configurator has `void accept(ServletContext servletContext, ServerContainer serverContainer) throws DeploymentException;`

We have SCI implementation specific code that executes after a successful `SCI.onStartup()` call.

Perhaps the method should be called `.postOnStartupConsumer(Consumer)` ?

Reply:
Also, the `Configurator` interface is for end users in embedded-jetty to perform actions when that SCI has completed and the object of note exists and is properly initialized.
Namely the `NativeWebSocketConfiguration` for jetty websockets, or the `ServerContainer` for JSR websockets.

Example of a user of Configurator ...

https://github.com/eclipse/jetty.project/blob/0119deeea13225ae1add5d405b04842559d7ad4d/jetty-websocket/javax-websocket-server-impl/src/test/java/org/eclipse/jetty/websocket/jsr356/server/BinaryStreamTest.java#L63-L70

Reply:
Changed consumer name to be more clear. added javadoc.